// @lumieducation/h5p-mongos3/build/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MongoLibraryStorage = exports.MongoS3LibraryStorage = exports.S3TemporaryFileStorage = exports.initMongo = exports.initS3 = exports.MongoS3ContentStorage = void 0;
var MongoS3ContentStorage_1 = __importDefault(require("./MongoS3ContentStorage"));
exports.MongoS3ContentStorage = MongoS3ContentStorage_1.default;
var initS3_1 = __importDefault(require("./initS3"));
exports.initS3 = initS3_1.default;
var initMongo_1 = __importDefault(require("./initMongo"));
exports.initMongo = initMongo_1.default;
var S3TemporaryFileStorage_1 = __importDefault(require("./S3TemporaryFileStorage"));
exports.S3TemporaryFileStorage = S3TemporaryFileStorage_1.default;
var MongoS3LibraryStorage_1 = __importDefault(require("./MongoS3LibraryStorage"));
exports.MongoS3LibraryStorage = MongoS3LibraryStorage_1.default;
var MongoLibraryStorage_1 = __importDefault(require("./MongoLibraryStorage"));
exports.MongoLibraryStorage = MongoLibraryStorage_1.default;
